# Importing libraries

In [37]:
import pandas as pd
import numpy as np
import os

# Importing data

In [38]:
# Path shortcut
path = r'/Users/lynnsey/Desktop/Bootcamp/Tasks/Immersion/Task 4/07-20-2023 Instacart Basket Analysis'

In [39]:
# Import products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [40]:
# Import wrangled orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Creating a mixed data frame to experiment with changing data type

In [41]:
# Create a dataframe
df_test = pd.DataFrame()

In [42]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [43]:
df_test.head

<bound method NDFrame.head of     mix
0     a
1     b
2     1
3  True>

In [44]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [45]:
# Change data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Finding missing values and removing them

In [46]:
# finding missing observations in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [47]:
# Creating a data frame containing only rows with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [48]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [49]:
df_prods.shape

(49693, 5)

In [50]:
# Create a new data frame without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [51]:
df_prods_clean.shape

(49677, 5)

# Checking for duplicates and removing them

In [52]:
# Checking for duplicates in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
df_prods_clean.shape

(49677, 5)

In [55]:
# Creating a new data frame without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [56]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting data as csv

In [57]:
#Exporting data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Questions

## Question 1

### Perform consistency checks on the products dataframe

In [58]:
# Looking at the products data frame for inconsistencies
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### The maximum price seems too high for a grocery product. I need to investigate this farther. 

In [59]:
# Sorting by price and showing rows with highest values
df_prods_clean_no_dups.sort_values('prices', ascending=False)

product_id                      product_name  aisle_id  department_id  \
33666       33664             2 % Reduced Fat  Milk        84             16   
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
19392       19391         Turkey Breast Tenderloins        49             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   
40490       40486                   Chicken Tenders        49             12   
...           ...                               ...       ...            ...   
11307       11307                  Goat Cheese Logs         2             16   
32805       32803                 Regular Deodorant        80             11   
46627       46623                   Pie Pans, Large        10             17   
37041       37037                       Redseedless        24              4   
21464       21463      Vanilla Bean Light Ice Cream        37              1   

        prices  
33666  99999.0  
21554  14900.0  
19392     25.0  
25580     25.0  
40490     25.0  
...        ...  
11307      1.0  
32805      1.0  
46627      1.0  
37041      1.0  
21464      1.0  

[49672 rows x 5 columns]

## Answer: 

### I can see that the 2% reduced fat milk price and the lowfat 2% milkfat cottage cheese is priced too high.  I would go back to the client and get an accurate price for each of these items at this point.  All other consistency checks have been undertaken above.  I would not change anything at the moment, until I have an answer from the client.  However, I would make a not that these two products are incorrectly priced until I have the correct price.

In [60]:
# Changing milk price from 99999.0 to 0.99 and cottage cheese price from 14900.0 to 1.49
df_prods_clean_no_dups = df_prods_clean_no_dups.replace ({"prices":{99999.0: 0.99, 14900.0:1.49 }})

In [61]:
# Checking to see if change went through
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==21553]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

       prices  
21554    1.49

In [62]:
# Checking to see if change went through
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    0.99

### I changed the decimal point to make the price of milk 0.99, which is about the price of a pint of milk.  I changed the decimal point on the cottage cheese as well to make it 1.49.

# All remaining questions refer to Order dataframe.

## Question 2 

### Run df.describe() function and share whether anything about the data looks off or should be investigated further.

In [63]:
# Looking at the orders data frame for inconsistencies
df_ords.describe().apply(lambda s: s.apply('{0:.1f}'.format))

Unnamed: 0   order_id    user_id order_number orders_day_of_week  \
count  3421083.0  3421083.0  3421083.0    3421083.0          3421083.0   
mean   1710541.0  1710542.0   102978.2         17.2                2.8   
std     987581.7   987581.7    59533.7         17.7                2.0   
min          0.0        1.0        1.0          1.0                0.0   
25%     855270.5   855271.5    51394.0          5.0                1.0   
50%    1710541.0  1710542.0   102689.0         11.0                3.0   
75%    2565811.5  2565812.5   154385.0         23.0                5.0   
max    3421082.0  3421083.0   206209.0        100.0                6.0   

      time_ordered days_since_prior_order  
count    3421083.0              3214874.0  
mean          13.5                   11.1  
std            4.2                    9.2  
min            0.0                    0.0  
25%           10.0                    4.0  
50%           13.0                    7.0  
75%           16.0                   15.0  
max           23.0                   30.0

### Answer: After looking at the descriptive statistics nothing jumps out at me as an obvious mistake.  The min and max values look correct for each column based on day of week, time and days since last order.

## Questions 3 & 4

### Check for mixed-type data and fix if found.

In [64]:
#Check for mixed type data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Answer: There are no mixed data types

## Question 5

### Run a check for missing values

In [65]:
# Check for missing observations
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
time_ordered                   0
days_since_prior_order    206209
dtype: int64

### Answer: There are 206,209 missing values in the 'days_since_prior_order' column.  The missing values could mean that the user has never placed an order before.

## Question 6

### Address the missing values using an appropriate method.

In [66]:
# Creating a data frame containing only rows with missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [67]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  time_ordered  days_since_prior_order  
0                         2             8                     NaN  
11                        2            11                     NaN  
26                        1            14                     NaN  
39                        6            11                     NaN  
45                        3            12                     NaN  
...                     ...           ...                     ...  
3420930                   4            12                     NaN  
3420934                   3            18                     NaN  
3421002                   6            19                     NaN  
3421019                   1            15                     NaN  
3421069                   3            11                     NaN  

[206209 rows x 8 columns]

### It looks like all the users with missing values in that column have only made 1 order.  Now I will check to see if the count of order_number with a value of 1 matches the count of missing values in the days_since_prior_order column.

In [68]:
#Checking to find the count of order_number with a value of 1.
df_ords.order_number.value_counts()[1]

206209

### Answer: I can see that the count or the order_number with a value of 1 and the count of the missing values for days_since_prior_order are both 206,209.  From this I can conclude that the missing values mean that no prior order has been placed.  I've chose to leave these missing values in as they are an important piece of information.  I do not want to change them to 0 because 0 would mean that someone could have placed their last order earlier in the day, not necessarily that they've never placed one.

## Questions 7 & 8

### Run a check for duplicates and explain why there might be duplicates; address the duplicates

In [69]:
# Checking for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [70]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, time_ordered, days_since_prior_order]
Index: []

### Answer: There are no duplicates

## Question 9

### Export the Orders data.  The product data was exported above and no changes have been made since then.

In [71]:
#Exporting the orders data
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [72]:
#Exporting the products data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [76]:
df_ords.shape

(3421083, 8)

In [77]:
# Dropped eval_set column as it had no useful information; I thought I did this earlier but it was still showing
df_ords.drop(columns = ['eval_set'])

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         time_ordered  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
...               ...                     ...  
3421078            18                    29.0  
3421079            10                    30.0  
3421080            12                    18.0  
3421081            12                     7.0  
3421082            14                    30.0  

[3421083 rows x 7 columns]

TypeError: 'Series' object is not callable